File: app/src/main/java/bin/mg/main/file/SymbolPanel.java
package bin.mg.main.file;

import android.animation.ValueAnimator;
import android.content.Context;
import android.graphics.Color;
import android.graphics.Typeface;
import android.graphics.drawable.GradientDrawable;
import android.graphics.drawable.StateListDrawable;
import android.util.AttributeSet;
import android.view.Gravity;
import android.view.MotionEvent;
import android.view.View;
import android.view.animation.DecelerateInterpolator;
import android.widget.LinearLayout;
import android.widget.TextView;

/**
 * A draggable symbol panel that sits above the soft keyboard.
 * Matches MT Manager's symbol panel:
 * - One row visible when collapsed
 * - Drag up to expand and reveal all symbol rows
 * - Drag down to collapse back to one row
 * - Smooth animations, state preserved
 */
public class SymbolPanel extends LinearLayout {

    public interface OnSymbolClickListener {
        void onSymbolClick(String symbol);
    }

    private OnSymbolClickListener mListener;

    private static final String[][] SYMBOL_ROWS = {
        {"→", "/", "+", "-", "*", "=", "<"},
        {">", "\"", "'", ";", "|", "\\", "-"},
        {"()", "[]", "{}", "..."}
    };

    private static final String[][] INSERT_ROWS = {
        {"\t", "/", "+", "-", "*", "=", "<"},
        {">", "\"", "'", ";", "|", "\\", "-"},
        {"()", "[]", "{}", "..."}
    };

    private LinearLayout rowsContainer;
    private boolean isExpanded = false;
    private boolean isAnimating = false;

    private static final int ROW_HEIGHT_DP = 42;
    private static final int DRAG_THRESHOLD_DP = 36;
    private static final long ANIMATION_DURATION = 180;

    private int handleTop;
    private int handleBottom;

    public SymbolPanel(Context context) {
        super(context);
        init(context);
    }

    public SymbolPanel(Context context, AttributeSet attrs) {
        super(context, attrs);
        init(context);
    }

    public SymbolPanel(Context context, AttributeSet attrs, int defStyleAttr) {
        super(context, attrs, defStyleAttr);
        init(context);
    }

    private void init(Context context) {
        setOrientation(VERTICAL);
        setLayoutParams(new LinearLayout.LayoutParams(
            LinearLayout.LayoutParams.MATCH_PARENT,
            LinearLayout.LayoutParams.WRAP_CONTENT
        ));

        float density = getResources().getDisplayMetrics().density;

        // Drag handle
        LinearLayout handleLayout = new LinearLayout(context);
        handleLayout.setOrientation(HORIZONTAL);
        handleLayout.setGravity(Gravity.CENTER_HORIZONTAL);
        handleLayout.setLayoutParams(new LinearLayout.LayoutParams(
            LinearLayout.LayoutParams.MATCH_PARENT,
            (int)(10 * density)
        ));
        handleLayout.setPadding(0, (int)(3 * density), 0, (int)(2 * density));

        View handle = new View(context);
        int handleW = (int)(32 * density);
        int handleH = (int)(3 * density);
        GradientDrawable handleBg = new GradientDrawable();
        handleBg.setShape(GradientDrawable.RECTANGLE);
        handleBg.setColor(0xFF555555);
        handleBg.setCornerRadius(handleH / 2f);
        handle.setBackground(handleBg);
        handleLayout.addView(handle);
        addView(handleLayout);

        rowsContainer = new LinearLayout(context);
        rowsContainer.setOrientation(VERTICAL);
        rowsContainer.setLayoutParams(new LinearLayout.LayoutParams(
            LinearLayout.LayoutParams.MATCH_PARENT,
            LinearLayout.LayoutParams.WRAP_CONTENT
        ));

        int textColor = 0xFFCCCCCC;
        int pressedColor = 0xFF444444;

        for (int i = 0; i < SYMBOL_ROWS.length; i++) {
            LinearLayout row = createSymbolRow(SYMBOL_ROWS[i], INSERT_ROWS[i], textColor, pressedColor, density);
            rowsContainer.addView(row);
            if (i > 0) {
                row.setVisibility(View.GONE);
            }
        }

        addView(rowsContainer);

        // Record handle area for drag detection
        handleLayout.post(() -> {
            int[] loc = new int[2];
            getLocationOnScreen(loc);
            handleTop = loc[1];
            handleBottom = handleTop + handleLayout.getHeight();
        });
    }

    private LinearLayout createSymbolRow(String[] symbols, String[] inserts, int textColor, int pressedColor, float density) {
        LinearLayout row = new LinearLayout(getContext());
        row.setOrientation(HORIZONTAL);
        row.setLayoutParams(new LinearLayout.LayoutParams(
            LinearLayout.LayoutParams.MATCH_PARENT,
            (int)(ROW_HEIGHT_DP * density)
        ));

        int dividerColor = 0xFF333333;

        for (int i = 0; i < symbols.length; i++) {
            final String insertText = inserts[i];

            TextView btn = new TextView(getContext());
            btn.setText(symbols[i]);
            btn.setTextColor(textColor);
            btn.setTextSize(16f);
            btn.setTypeface(Typeface.MONOSPACE);
            btn.setGravity(Gravity.CENTER);
            btn.setLayoutParams(new LinearLayout.LayoutParams(
                0,
                LinearLayout.LayoutParams.MATCH_PARENT,
                1f
            ));
            btn.setClickable(true);
            btn.setFocusable(true);

            // Pressed state background
            StateListDrawable bg = new StateListDrawable();
            GradientDrawable pressed = new GradientDrawable();
            pressed.setShape(GradientDrawable.RECTANGLE);
            pressed.setColor(pressedColor);
            GradientDrawable normal = new GradientDrawable();
            normal.setShape(GradientDrawable.RECTANGLE);
            normal.setColor(Color.TRANSPARENT);
            bg.addState(new int[]{android.R.attr.state_pressed}, pressed);
            bg.addState(new int[]{}, normal);
            btn.setBackground(bg);

            btn.setMinWidth(0);
            btn.setMinimumWidth(0);
            btn.setPadding(0, 0, 0, 0);

            btn.setOnClickListener(v -> {
                if (mListener != null) {
                    mListener.onSymbolClick(insertText);
                }
            });

            row.addView(btn);

            if (i < symbols.length - 1) {
                View divider = new View(getContext());
                divider.setLayoutParams(new LinearLayout.LayoutParams(
                    (int)(1 * density),
                    LinearLayout.LayoutParams.MATCH_PARENT
                ));
                divider.setBackgroundColor(dividerColor);
                row.addView(divider);
            }
        }

        return row;
    }

    public void setOnSymbolClickListener(OnSymbolClickListener listener) {
        mListener = listener;
    }

    public boolean isExpanded() {
        return isExpanded;
    }

    public void expand() {
        if (isExpanded || isAnimating) return;
        isExpanded = true;
        updateVisibility(true);
    }

    public void collapse() {
        if (!isExpanded || isAnimating) return;
        isExpanded = false;
        updateVisibility(false);
    }

    public void toggle() {
        if (isExpanded) collapse();
        else expand();
    }

    private void updateVisibility(boolean showAll) {
        isAnimating = true;
        int childCount = rowsContainer.getChildCount();

        for (int i = 1; i < childCount; i++) {
            final View row = rowsContainer.getChildAt(i);
            if (showAll) {
                row.setVisibility(View.VISIBLE);
                row.setAlpha(0f);
                row.animate()
                    .alpha(1f)
                    .setDuration(ANIMATION_DURATION)
                    .setInterpolator(new DecelerateInterpolator())
                    .withEndAction(() -> {
                        if (i == childCount - 1) isAnimating = false;
                    })
                    .start();
            } else {
                final int index = i;
                row.animate()
                    .alpha(0f)
                    .setDuration(ANIMATION_DURATION)
                    .setInterpolator(new DecelerateInterpolator())
                    .withEndAction(() -> {
                        row.setVisibility(View.GONE);
                        row.setAlpha(1f);
                        if (index == childCount - 1) isAnimating = false;
                    })
                    .start();
            }
        }

        if (childCount <= 1) {
            isAnimating = false;
        }
    }

    @Override
    public boolean onInterceptTouchEvent(MotionEvent ev) {
        // Only intercept drags that start on the handle area
        if (ev.getAction() == MotionEvent.ACTION_DOWN) {
            int[] loc = new int[2];
            getLocationOnScreen(loc);
            float rawY = ev.getRawY();
            // Check if touch is in the top 14dp (handle area)
            float density = getResources().getDisplayMetrics().density;
            float handleAreaBottom = loc[1] + 14 * density;
            if (rawY >= loc[1] && rawY <= handleAreaBottom) {
                dragStartRawY = ev.getRawY();
                isDragging = false;
                return false; // Don't intercept yet, wait for movement
            }
        }
        return false;
    }

    private float dragStartRawY;
    private boolean isDragging;

    @Override
    public boolean onTouchEvent(MotionEvent event) {
        float density = getResources().getDisplayMetrics().density;
        float threshold = DRAG_THRESHOLD_DP * density;

        switch (event.getAction()) {
            case MotionEvent.ACTION_DOWN:
                dragStartRawY = event.getRawY();
                isDragging = false;
                return true;

            case MotionEvent.ACTION_MOVE:
                float dy = dragStartRawY - event.getRawY();
                if (Math.abs(dy) > threshold) {
                    isDragging = true;
                }
                return true;

            case MotionEvent.ACTION_UP:
                if (isDragging) {
                    float totalDy = dragStartRawY - event.getRawY();
                    if (totalDy > threshold && !isExpanded) {
                        expand();
                    } else if (totalDy < -threshold && isExpanded) {
                        collapse();
                    }
                }
                isDragging = false;
                return true;
        }
        return false;
    }
}

File: app/src/main/java/bin/mg/main/file/SymbolPanel.kt
package bin.mg.main.file

import android.content.Context
import android.content.SharedPreferences
import android.util.AttributeSet
import android.view.Gravity
import android.view.LayoutInflater
import android.view.View
import android.widget.LinearLayout
import android.widget.TextView
import androidx.core.view.ViewCompat
import androidx.core.view.WindowInsetsCompat
import com.google.android.material.bottomsheet.BottomSheetBehavior
import com.google.android.material.bottomsheet.BottomSheetDragHandleView
import com.google.android.material.shape.MaterialShapeDrawable
import com.google.android.material.shape.ShapeAppearanceModel

class SymbolPanel @JvmOverloads constructor(
    context: Context,
    attrs: AttributeSet? = null,
    defStyleAttr: Int = 0
) : LinearLayout(context, attrs, defStyleAttr) {

    interface OnSymbolClickListener {
        fun onSymbolClick(symbol: String)
    }

    private var symbolClickListener: OnSymbolClickListener? = null
    private lateinit var bottomSheetBehavior: BottomSheetBehavior<LinearLayout>
    private lateinit var dragHandle: BottomSheetDragHandleView
    private lateinit var rowsContainer: LinearLayout
    
    private companion object {
        private val SYMBOL_ROWS = arrayOf(
            arrayOf("→", "/", "+", "-", "*", "=", "<"),
            arrayOf(">", "\"", "'", ";", "|", "\\", "_"),
            arrayOf("()", "[]", "{}", "...")
        )
        
        private val INSERT_ROWS = arrayOf(
            arrayOf(null, "/", "+", "-", "*", "=", "<"),
            arrayOf(">", "\"", "'", ";", "|", "\\", "_"),
            arrayOf("()", "[]", "{}", "...")
        )
        
        private const val ROW_HEIGHT_DP = 48
        private const val PEEK_HEIGHT_DP = 56
        private const val EXPANDED_RATIO = 0.4f
    }

    init {
        orientation = VERTICAL
        initView()
        setupBottomSheetBehavior()
        setupKeyboardInsetsListener()
    }

    private fun initView() {
        // Create drag handle
        dragHandle = BottomSheetDragHandleView(context).apply {
            layoutParams = LayoutParams(
                LayoutParams.MATCH_PARENT,
                LayoutParams.WRAP_CONTENT
            )
            setGravity(Gravity.CENTER_HORIZONTAL)
        }
        addView(dragHandle)

        // Create rows container
        rowsContainer = LinearLayout(context).apply {
            orientation = VERTICAL
            layoutParams = LayoutParams(
                LayoutParams.MATCH_PARENT,
                LayoutParams.WRAP_CONTENT
            )
        }
        addView(rowsContainer)

        // Create symbol rows
        val density = resources.displayMetrics.density
        SYMBOL_ROWS.forEachIndexed { rowIndex, _ ->
            val row = createSymbolRow(rowIndex, density)
            rowsContainer.addView(row)
            
            // Hide rows beyond first when collapsed
            if (rowIndex > 0) {
                row.visibility = GONE
            }
        }
    }

    private fun createSymbolRow(rowIndex: Int, density: Float): LinearLayout {
        val labels = SYMBOL_ROWS[rowIndex]
        val inserts = INSERT_ROWS[rowIndex]
        
        return LinearLayout(context).apply {
            orientation = HORIZONTAL
            layoutParams = LayoutParams(
                LayoutParams.MATCH_PARENT,
                (ROW_HEIGHT_DP * density).toInt()
            )
            
            labels.forEachIndexed { colIndex, label ->
                val button = TextView(context).apply {
                    text = label
                    setTextColor(0xFFCCCCCC.toInt())
                    textSize = 15f
                    typeface = android.graphics.Typeface.MONOSPACE
                    gravity = Gravity.CENTER
                    layoutParams = LayoutParams(
                        0,
                        LayoutParams.MATCH_PARENT,
                        1f
                    )
                    isClickable = true
                    isFocusable = true
                    
                    // Material design ripple effect
                    ViewCompat.setBackground(this, createRippleDrawable())
                    
                    setOnClickListener {
                        val symbol = when {
                            rowIndex == 0 && colIndex == 0 -> buildTabString()
                            else -> inserts[colIndex] ?: return@setOnClickListener
                        }
                        symbolClickListener?.onSymbolClick(symbol)
                    }
                }
                addView(button)
            }
        }
    }

    private fun createRippleDrawable() = MaterialShapeDrawable(
        ShapeAppearanceModel.builder()
            .setAllCorners(com.google.android.material.shape.CornerFamily.ROUNDED, 0f)
            .build()
    ).apply {
        fillColor = android.content.res.ColorStateList.valueOf(android.graphics.Color.TRANSPARENT)
        setTint(0x33FFFFFF.toInt())
    }

    private fun buildTabString(): String {
        val prefs = context.getSharedPreferences("editor_prefs", Context.MODE_PRIVATE)
        val useTabs = prefs.getBoolean("use_tabs", false)
        
        return if (useTabs) {
            "\t"
        } else {
            val tabSize = prefs.getInt("tab_size", 4)
            " ".repeat(tabSize)
        }
    }

    private fun setupBottomSheetBehavior() {
        bottomSheetBehavior = BottomSheetBehavior.from(this).apply {
            peekHeight = (PEEK_HEIGHT_DP * resources.displayMetrics.density).toInt()
            expandedOffset = 0
            isHideable = false
            isFitToContents = false
            halfExpandedRatio = EXPANDED_RATIO
            
            addBottomSheetCallback(object : BottomSheetBehavior.BottomSheetCallback() {
                override fun onStateChanged(bottomSheet: View, newState: Int) {
                    when (newState) {
                        BottomSheetBehavior.STATE_EXPANDED,
                        BottomSheetBehavior.STATE_HALF_EXPANDED -> showAllRows()
                        BottomSheetBehavior.STATE_COLLAPSED -> showFirstRowOnly()
                    }
                }
                
                override fun onSlide(bottomSheet: View, slideOffset: Float) {
                    // slideOffset: 0 = collapsed, 1 = expanded
                    updateRowsVisibility(slideOffset)
                }
            })
        }
    }

    private fun updateRowsVisibility(slideOffset: Float) {
        rowsContainer.children.forEachIndexed { index, view ->
            if (index > 0) { // Skip first row (index 0)
                val targetAlpha = slideOffset.coerceIn(0f, 1f)
                view.alpha = targetAlpha
                view.visibility = if (targetAlpha > 0.01f) VISIBLE else GONE
            }
        }
    }

    private fun showAllRows() {
        rowsContainer.children.forEach { it.visibility = VISIBLE }
    }

    private fun showFirstRowOnly() {
        rowsContainer.children.forEachIndexed { index, view ->
            view.visibility = if (index == 0) VISIBLE else GONE
        }
    }

    private fun setupKeyboardInsetsListener() {
        ViewCompat.setOnApplyWindowInsetsListener(this) { _, insets ->
            val imeInsets = insets.getInsets(WindowInsetsCompat.Type.ime())
            val systemBarsInsets = insets.getInsets(WindowInsetsCompat.Type.systemBars())
            
            // Position the panel above the keyboard
            val bottomPadding = imeInsets.bottom - systemBarsInsets.bottom
            if (bottomPadding > 0) {
                setPadding(paddingLeft, paddingTop, paddingRight, 0)
                bottomSheetBehavior.peekHeight = (PEEK_HEIGHT_DP * resources.displayMetrics.density).toInt()
            } else {
                setPadding(paddingLeft, paddingTop, paddingRight, systemBarsInsets.bottom)
            }
            
            insets
        }
    }

    fun setOnSymbolClickListener(listener: OnSymbolClickListener) {
        symbolClickListener = listener
    }

    fun expand() {
        bottomSheetBehavior.state = BottomSheetBehavior.STATE_EXPANDED
    }

    fun collapse() {
        bottomSheetBehavior.state = BottomSheetBehavior.STATE_COLLAPSED
    }

    fun toggle() {
        when (bottomSheetBehavior.state) {
            BottomSheetBehavior.STATE_COLLAPSED -> expand()
            else -> collapse()
        }
    }

    fun isExpanded(): Boolean {
        return bottomSheetBehavior.state == BottomSheetBehavior.STATE_EXPANDED ||
               bottomSheetBehavior.state == BottomSheetBehavior.STATE_HALF_EXPANDED
    }
}